### DISCUSSION TITLE
Modbus Soft PLC Using modbus TCP Server and Modbus TCP Client

### DISCUSSION THREAD

**Discussion by bambosd** [pymodbus-dev/pymodbus#2281] [General]
Hello everyone,

i'm checking out if is possible and how we can have a soft modbus PLC using the current pymodbus library, and all the functionality of python. 

In theory, a modbus PLC have inputs by 2 methods. 
a) having a value written on it's modbus TCP Server memory area from another modbus TCP client. (coil or register int16 or float32)
b) having a value read by its self from another modbus server and keep it on it's own registers , or internal variable.

Then there is process, calculations, evaluations, control loops etc.... which python is overpowered of course compared to a PLC.

And then there is outputs. Outputs can be write coil to another device or write register to another device (int16 or float32).

If we can structure all above functionality in functions without interrupt each other and make somehow multitask all the jobs, then is good enough. Is there any included utilities in pymodbus to facilitate this ? can this be achieved with async functions or something ?? 

any suggestions and comments appreciated. (i'm going to testing)

**Comment by janiversen** [comment]
Please see our examples, pymodbus server seems to offer what you require.

**Comment by janiversen** [comment]
Maybe the simultator would be a better fit.

**Comment by bambosd** [comment]
Is there any way to get values in the modbus tcp server internally? I had online a pymodbus tcp server a while ago and it seems that once is running, there is no way to get values internally.
So a possible solution maybe is to have a client and write on its self. 127.0.0.1.
What do you think?

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by janiversen** [reply]
??? Did you look at the simulator it allows you to attach functions that do whatever you want.

We also have a server example where you set the values from an outside process

So please have a look at our examples, with them all your demands are covered.

**Reply by janiversen** [reply]
Simulator is not more than an enhanced datastore + server + optional http, as documented.

**Comment by bambosd** [comment]
Thanks a lot ! I will work on those two options you suggested.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
